### PR TITLE
Upgrade dependencies

### LINKED ISSUE #3115: Update to the latest compatible versions of all dependencies
## Motivation

Zebra isn't using the latest versions of its dependencies, because we skipped some dependency updates during the NU5 sprints.

We should update dependencies, to get as many bug and security fixes as possible.

**Note: we've been discussing in Discord how to split this ticket into smaller parts**

### Tasks

- Run `cargo update` to update to the latest versions of each dependency
- Revert any incompatible version updates (but try to find the latest compatible version)
- Revert any dependency changes that cause a lot of duplicates (particularly with tower, tracing, and orchard)
- Update the duplicate dependency checks in `deny.toml`

### PR DESCRIPTION
**Pull request by upbqdn**
## Motivation

Zebra should use the newest versions of its dependencies.

## Solution

I used `cargo upgrade --workspace` to upgrade all dependencies. This caused a lot of errors during the compilation. Here's a list containing the diffs of the problematic bumps:

-pin-project = "0.4.29"
+pin-project = "1.0.10"

-bitvec = "0.22"
+bitvec = "1.0.0"

-halo2 = "=0.1.0-beta.1"
+halo2 = "0.1.0-beta.2"

-ripemd160 = "0.9"
+ripemd160 = "0.10.0"

-secp256k1 = { version = "0.21.2", features = ["serde"] }
+secp256k1 = { version = "0.21.3", features = ["serde"] }

-sha2 = { version = "0.9.9", features=["compress"] }
+sha2 = { version = "0.10.2", features = ["compress"] }

-proptest = { version = "0.10.1", optional = true }
+proptest = { version = "1.0.0", optional = true }

-proptest = "0.10.1"
+proptest = "1.0.0"

-rand07 = { package = "rand", version = "0.7" }
+rand07 = { package = "rand", version = "0.8.5" }

-metrics = "0.17.1"
+metrics = "0.18.0"

-arti-client = { version = "0.0.2", optional = true }
+arti-client = { version = "0.0.4", optional = true }

-tor-rtcompat  = { version = "0.0.2", optional = true }
+tor-rtcompat  = { version = "0.0.4", optional = true }

-color-eyre = "0.5.11"
+color-eyre = "0.6.0"

-tracing-subscriber = "0.2.25"
+tracing-subscriber = "0.3.9"

-tracing-error = "0.1.2"
+tracing-error = "0.2.0"

-abscissa_core = "0.5"
+abscissa_core = "0.6.0"

-gumdrop = "0.7"
+gumdrop = "0.8.0"

-color-eyre = { version = "0.5.11", features = ["issue-url"] }
+color-eyre = { version = "0.6.0", features = ["issue-url"] }

-tracing-flame = "0.1.0"
+tracing-flame = "0.2.0"

-tracing-journald = "0.1.0"
+tracing-journald = "0.2.3"

-tracing-subscriber = { version = "0.2.25", features = ["tracing-log"] }
+tracing-subscriber = { version = "0.3.9", features = ["tracing-log"] }

-metrics-exporter-prometheus = "0.7.0"
+metrics-exporter-prometheus = "0.8.0"

-sentry-tracing = "0.23.0"
+sentry-tracing = "0.24.3"

-abscissa_core = { version = "0.5", features = ["testing"] }
+abscissa_core = { version = "0.6.0", features = ["testing"] }

--- 

Every other dependency was upgraded to the newest version. The way I produced the list above was that I ran `cargo upgrade --workspace`, and then I was manually identifying the bumps that were causing errors and reverting them until I got a clean build and all tests were passing.

Will close #3115.

## Review

Anyone can review.

### CODE REVIEW

**Comment by upbqdn** [comment]
I wasn't looking for the last compatible version of each dependency because I didn't find a quick way to automate this process. I think it would be tedious to identify such versions by hand, and we should switch to the newest versions anyway at some point. Is the latter assumption correct?

I'm not sure what's meant by this statement from the ticket:
> Revert any dependency changes that cause a lot of duplicates (particularly with tower, tracing, and orchard)

In particular, how do I identify the duplicates?

I labeled the PR as a draft for now.

**Comment by teor2345** [comment]
> I wasn't looking for the last compatible version of each dependency because I didn't find a quick way to automate this process. I think it would be tedious to identify such versions by hand, and we should switch to the newest versions anyway at some point. Is the latter assumption correct?

Yep, this seems like a lot of work.

Let's delay this decision until we have time to work on tricky dependency upgrades. (We might also decide to specify the minimum compatible version in library crates, and the latest in the lock file for binaries. Because this makes libraries easier for non-zebra projects to use.)

> I'm not sure what's meant by this statement from the ticket:
> 
> > Revert any dependency changes that cause a lot of duplicates (particularly with tower, tracing, and orchard)
> 
> In particular, how do I identify the duplicates?

Use `cargo deny check bans` or this CI job:
https://github.com/ZcashFoundation/zebra/runs/5311535602?check_suite_focus=true

Then for each duplicate, work out if the crate: 
- needs a single version to function correctly: metrics, tracing, sentry, tokio
- needs a single version for consensus compatibility: maybe blake2 ?
- is small and can be ignored in deny.toml
- is large and should be de-duplicated

We try to de-duplicate crates where possible. Usually it just takes a version bump of another crate, or a version revert.

**Comment by teor2345** [comment]
I think we're leaving the blake2 upgrades until #2982, with the rest of the testnet cryptography upgrades.

**Comment by upbqdn** [comment]
I've reverted the following dependencies because of duplicates.

* Needs a single version for consensus compatibility: 
  * -blake2s_simd = "1.0.0"
    +blake2s_simd = "0.5.11"

* Needs a single version to function correctly:
  * -metrics = "0.18.0"
    +metrics = "0.17.1"
  * -tracing-error = "0.2.0"
     +tracing-error = "0.1.2"
  * -tracing-subscriber = "0.3.9"
     +tracing-subscriber = "0.2.25"
  * -sentry = { version = "0.24.3", default-features = false, features = ["backtrace", "contexts", "reqwest", "rustls"] }
    +sentry = { version = "0.23.0", default-features = false, features = ["backtrace", "contexts", "reqwest", "rustls"] }

**Comment by upbqdn** [comment]
There are still duplicates. They seem to be caused by `arti-client v0.0.2` and `abscissa`, but I didn't change the version of these crates. I'm trying to figure out why these duplicates occur, but no luck so far.

**Comment by upbqdn** [comment]
I noticed I'm not able to change `abscissa` to `0.5.1` or `0.5.0`. ` cargo deny --workspace --all-features check bans
` still outputs `abscissa_core v0.5.2`.

**Comment by teor2345** [comment]
> I noticed I'm not able to change `abscissa` to `0.5.1` or `0.5.0`. `cargo deny --workspace --all-features check bans` still outputs `abscissa_core v0.5.2`.

It looks like we're almost there, I'll see what I can do to help here.

We might just need to ignore some duplicates.

**Comment by teor2345** [comment]
@upbqdn what do you think of the duplicate fixes in #3716?

Feel free to merge it into this PR.

File: zebra-network/proptest-regressions/peer_set/candidate_set/tests/prop.txt [hunk 1]
```diff
@@ -6,3 +6,5 @@
...
+cc 87f0ffdadade0b7633d1442af5cef2115dd8bd7e6356a170f2c100bd18d1dce0 # shrinks to next_peer_attempts = 2
+cc 4b5a3e953255ede0bb3818dd3e4cbf7598fe9b4e59cf10253c62edfae550c58e # shrinks to next_peer_attempts = 7
```
Comment (teor2345) [inline zebra-network/proptest-regressions/peer_set/candidate_set/tests/prop.txt]: These test cases seem unrelated, can we remove them?

Comment (upbqdn) [reply zebra-network/proptest-regressions/peer_set/candidate_set/tests/prop.txt]: Done.

**Review by teor2345 — DISMISSED** [review]
Looks great, thanks!

**Comment by teor2345** [comment]
I opened PR #3726 for one of the CI failures, and the other is ticket #3712.

**Review by teor2345 — APPROVED** [review]
I'm going to merge this PR manually, because it's blocking other PRs like #3707.

And sometimes when we merge a new PR, there are conflicts with this PR.

I've tested it locally with PR #3755, and all the tests pass. (I didn't test the cached state or full sync.)

**Comment by teor2345** [comment]
I opened this PR for the integration test failure:
- #3758 

And this one for the rest of the test hangs:
- #3755
